osquery/tables/system/windows/windows_security_center.cpp

```diff
@@ -0,0 +1,125 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#include <windows.h>
+#include <wscapi.h>
+
+#include <osquery/core.h>
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+#include <osquery/tables.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/map_take.h>
+
+namespace osquery {
+namespace tables {
+
+const auto kSecurityProviderStates = std::unordered_map<int, std::string>{
+    {WSC_SECURITY_PROVIDER_HEALTH_NOTMONITORED, "Not Monitored"},
+    {WSC_SECURITY_PROVIDER_HEALTH_GOOD, "Good"},
+    {WSC_SECURITY_PROVIDER_HEALTH_POOR, "Poor"},
+    {WSC_SECURITY_PROVIDER_HEALTH_SNOOZE, "Snoozed"},
+};
+
+std::string resolveProductHealthOrError(int productName) {
+  // Attempt a runtime link to the DLL containing these functions,
+  // since linking the library was causing a crash on some Windows
+  // machines (like the CI server).
+  typedef HRESULT(WINAPI * pWscGetSecurityProviderHealth)(
+      _In_ DWORD Providers, _Out_ PWSC_SECURITY_PROVIDER_HEALTH);
+  pWscGetSecurityProviderHealth WscGetSecurityProviderHealth;
+  static HMODULE hDLL = LoadLibrary(TEXT("wscapi.dll"));
```

Comment:
I don’t think we should be using TEXT like this. It was only intended for virtual table data.

Reply:
Are you sure that LoadLibrary is not called multiple times? The intention is that this static name is only initialized once, let’s make sure that is true, if it wasn’t I assume a leak would occur?

Reply:
You are right, this is called multiple times (which I will fix). Looking at the MSDN docs calling LoadLibrary multiple times on the same DLL will simply increase the reference count.

From https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya?redirectedfrom=MSDN

> The system maintains a per-process reference count on all loaded modules. Calling LoadLibrary increments the reference count. Calling the FreeLibrary or FreeLibraryAndExitThread function decrements the reference count. The system unloads a module when its reference count reaches zero or when the process terminates (regardless of the reference count).

That being said, we can simply check if hDLL is already populated, instead of blindly incrementing the reference count over and over again. We should also probably unload the library once we do not need it anymore (at the end of the table's execution).

Reply:
My reading of the C++ standard indicates that `static` local variables are initialized exactly once (unless an exception occurs during initialization).

https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

Do you interpret this differently or see evidence that this is not the case? I would very much like to know if my understanding is incorrect.

Reply:
@zwass That's correct, `static` guarantees that the variable is initialized only once and in a thread-safe way too (since C++11), so the `LoadLibrary` function will only be called once.

Reply:
That’s how I interpret it (only once), and that’s generally been my experience, but I asked to make sure. For example, can you test using this table multiple times and then inspect the processes handles to make sure nothing is leaking.

Reply:
Ok I validated manually with the sysinternals tool handle and process explorer this is how it was behaving (not loading the dll multiple times). I think this is good to go now (I took out the TEXT function as it was extraneous)